src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -580,18 +577,51 @@ namespace stan {
         return;
       }
 
+       // check argument qualifiers
+      if (!decl.body_.is_no_op_statement()) {
+        function_signature_t decl_sig =
+          function_signatures::instance().get_definition(decl.name_, sig);
+        if (!decl_sig.first.is_ill_formed()) {
+          if (decl_sig.second.size() != arg_types.size()) {
+            error_msgs << "Declaration, definition mismatch for function "
+                       << decl.name_
+                       << ": declaration has " << arg_types.size()
+                       << " arguments, definition has " << arg_types.size()
+                       << "arguments." << std::endl;
+            pass = false;
+            return;
+          }
+          for (size_t i = 0; i < arg_types.size(); ++i) {
+            if (decl_sig.second[i].expr_type_ != arg_types[i].expr_type_
+                || decl_sig.second[i].data_only_ != arg_types[i].data_only_) {
+              error_msgs << "Declaration, definition mismatch for function "
+                         << decl.name_ << " argument " << (i + 1)
+                         << ": argument declared as "
+                         << (decl_sig.second[i].data_only_ ? "" : "data ")
+                         << decl_sig.second[i].expr_type_
+                         << ", defined as "
+                         << (arg_types[i].data_only_ ? "" : "data ")
+                         << arg_types[i].expr_type_ << "." <<  std::endl;
+              pass = false;
+              return;
+            }
+          }
+        }
+      }
 
-      if (ends_with("_lpdf", decl.name_) && arg_types[0].base_type_ == INT_T) {
+      if (ends_with("_lpdf", decl.name_)
+          && arg_types[0].expr_type_.base_type_ == INT_T) {
         error_msgs << "Parse Error.  Probability density functions require"
                    << " real variates (first argument)."
-                   << " Found type = " << arg_types[0] << std::endl;
+                   << " Found type = " << arg_types[0].expr_type_ << std::endl;
```

Comment:
I think the right thing to do here is add the `operator<<` for printing the new argument types.  That'll let you print the "data" qualification where it appears.

Reply:
done.